### PR TITLE
Ursa migration

### PR DESCRIPTION
**Pull request by mikelodder7**

### CODE REVIEW

File: libindy/src/commands/anoncreds/prover.rs [hunk 1]
```diff
@@ -279,7 +279,7 @@ impl ProverCommandExecutor {
 
         let credential_request_metadata = CredentialRequestMetadata {
             master_secret_blinding_data: ms_blinding_data,
-            nonce: credential_request.nonce.clone()?,
+            nonce: credential_request.nonce.try_clone()?,
```
Comment (kdenhartog) [inline libindy/src/commands/anoncreds/prover.rs]: leaving a note for myself to compare the differences between **clone()** and **try_clone()**

File: libindy/src/commands/anoncreds/tails.rs [hunk 2]
```diff
@@ -80,7 +80,7 @@ pub fn store_tails_from_generator(service: Rc<BlobStorageService>,
     let version = vec![0u8, TAILS_BLOB_TAG_SZ];
     service.append(blob_handle, version.as_slice())?;
 
-    while let Some(tail) = rtg.next()? {
+    while let Some(tail) = rtg.try_next()? {
```
Comment (kdenhartog) [inline libindy/src/commands/anoncreds/tails.rs]: compare differences between **next()** and **try_next()**

File: libindy/src/services/anoncreds/prover.rs [hunk 3]
```diff
@@ -373,7 +373,22 @@ impl Prover {
...
                     .to_indy(IndyErrorKind::InvalidStructure, format!("Credential attribute value \"{:?}\" is invalid", attribute_value))?;
                 Ok(attribute_value >= predicate.p_value)
-            }
+            },
+            PredicateTypes::GT => {
+                 let attribute_value = attribute_value.parse::<i32>()
+                    .to_indy(IndyErrorKind::InvalidStructure, format!("Credential attribute value \"{:?}\" is invalid", attribute_value))?;
+                 Ok(attribute_value > predicate.p_value)
+             },
+             PredicateTypes::LE => {
```
Comment (kdenhartog) [inline libindy/src/services/anoncreds/prover.rs]: Are we leaving off "Greater than or Equal to" PredicateType for a reason?

File: libindy/src/services/pool/networker.rs [hunk 4]
```diff
@@ -182,7 +183,8 @@ impl PoolConnection {
     fn new(mut nodes: Vec<RemoteNode>, active_timeout: i64, preordered_nodes: Vec<String>) -> Self {
         trace!("PoolConnection::new: from nodes {:?}", nodes);
 
-        thread_rng().shuffle(nodes.as_mut());
+        nodes.shuffle(&mut thread_rng());
+//        thread_rng().shuffle(nodes.as_mut());
```
Comment (kdenhartog) [inline libindy/src/services/pool/networker.rs]: If this is not being used, let's delete this line.

File: libindy/src/services/wallet/language.rs [hunk 5]
```diff
@@ -300,15 +300,16 @@ mod tests {
 
     use super::*;
     use self::rand::{thread_rng, Rng};
+    use rand::distributions::{Alphanumeric, Standard};
     use std::hash::Hash;
     use std::collections::HashSet;
 
     fn _random_vector(len: usize) -> Vec<u8> {
-        thread_rng().gen_iter().take(len).collect()
+        thread_rng().sample_iter(&Standard).take(len).collect()
```
Comment (kdenhartog) [inline libindy/src/services/wallet/language.rs]: note to compare differences between **gen_iter()** and **sample_iter()**

File: libindy/tests/anoncreds.rs [hunk 6]
```diff
@@ -3539,14 +3540,13 @@ mod medium_cases {
                "version":"0.1",
                "requested_attributes": json!({}),
                "requested_predicates": json!({
-                   "predicate1_referent": json!({ "name":"age", "p_type":"<=", "p_value":18 }),
+                   "predicate1_referent": json!({ "name":"age", "p_type":"!=", "p_value":18 }),
```
Comment (kdenhartog) [inline libindy/tests/anoncreds.rs]: Where is the change in p_type described in the API documentation? Can you add it, and additional tests that would support the new predicate types that are being added?

File: libindy/src/services/anoncreds/prover.rs (hunk 3 above)
Comment (kdenhartog) [reply libindy/src/services/anoncreds/prover.rs]: This functionality isn't being exposed at the API layer, so adding this shouldn't be a reason to block this PR. It's prep work for when we decide to expose this to the API, at which point we should add tests.

File: libindy/tests/anoncreds.rs (hunk 6 above)
Comment (kdenhartog) [reply libindy/tests/anoncreds.rs]: The change in p_types are not necessary at this point because the additional p_types aren't exposed at the API layer yet. This change is fine because it's a test to make sure the invalid p_type is properly throwing an error.

File: libindy/src/commands/anoncreds/prover.rs (hunk 1 above)
Comment (kdenhartog) [reply libindy/src/commands/anoncreds/prover.rs]: From what I can tell, this **try_clone()** function is performing a different operation than the **clone()** trait which was originally used. **try_clone()** appears to be generating a new BigNum. I got to this by looking at this line here: https://github.com/hyperledger/ursa/blob/0d03b1f071af652e3fa849241546dd0867ad3d51/libursa/src/bn/openssl.rs#L404

which then got me to look at the BigNum impl documentation in SSL, which says "Creates a new `BigNum` from an unsigned, big-endian encoded number of arbitrary length."

https://docs.rs/crate/openssl/0.10.20/source/src/bn.rs

Should we be generating a new nonce for this or do we want to be cloning an already created nonce?

File: libindy/src/commands/anoncreds/tails.rs (hunk 2 above)
Comment (kdenhartog) [reply libindy/src/commands/anoncreds/tails.rs]: The functionality appears to do the same things. However, this new function looks like it's generating a new Tails object each time and returning the new object rather than iterating on the same object. I was not able to find the impl of the **next()** trait in the std library to verify they do it differently. My only concern with this would be that it's not as memory efficient, which given the size of the Tails object may be a problem. Was there a reason for switching this?

File: libindy/src/services/wallet/language.rs (hunk 5 above)
Comment (kdenhartog) [reply libindy/src/services/wallet/language.rs]: From what I can tell, **sample_iter()** is not random. Is this intentional to constrain the test? In any case this shouldn't introduce any problems given it is used to generate randomness in tests.

File: libindy/src/commands/anoncreds/prover.rs (hunk 1 above)
Comment (kdenhartog) [reply libindy/src/commands/anoncreds/prover.rs]: No issues here. **clone()** is not a reference to a trait, but rather to an indy_crypto function found here. This is just a function rename to create more idiomatic rust. See the original function called here: https://github.com/hyperledger/indy-crypto/blob/1675e29a2a5949b44899553d3d128335cf7a61b3/libindy-crypto/src/bn/openssl.rs#L404

File: libindy/src/commands/anoncreds/tails.rs (hunk 2 above)
Comment (kdenhartog) [reply libindy/src/commands/anoncreds/tails.rs]: No problems here. This is not referring to the trait **next()** which is performed on an iterator, but rather to the **next()** function which can be found in Indy_crypto here: https://github.com/hyperledger/indy-crypto/blob/ca98135687a0674722f94836a2359ac5fd629b54/libindy-crypto/src/cl/mod.rs#L504

File: libindy/tests/crypto.rs [hunk 7]
```diff
@@ -195,7 +195,7 @@ mod high_cases {
         use super::*;
 
         #[test]
-        fn indy_crypto_sign_works() {
+        fn ursa_sign_works() {
```
Comment (jovfer) [inline libindy/tests/crypto.rs]: Seems like tests names in this file was auto-replaced from composite prefixes `indy` `crypto` to `ursa`.
Please revert these changes.

File: libindy/src/services/pool/networker.rs (hunk 4 above)
Comment (jovfer) [reply libindy/src/services/pool/networker.rs]: +1
